tsfc/finatinterface.py

```diff
@@ -179,9 +191,19 @@ def convert_tensorproductelement(element, **kwargs):
     cell = element.cell()
     if type(cell) is not ufl.TensorProductCell:
         raise ValueError("TensorProductElement not on TensorProductCell?")
-    elements, deps = zip(*[_create_element(elem, **kwargs)
-                           for elem in element.sub_elements()])
-    return finat.TensorProductElement(elements), set.union(*deps)
+    shift_axes = kwargs["shift_axes"]
+    dim_offset = 0
+
+    elements = []
+    deps = set()
+    for elem in element.sub_elements():
+        kwargs["shift_axes"] = shift_axes + dim_offset
+        dim_offset += elem.cell().topological_dimension()
+
+        finat_elem, ds = _create_element(elem, **kwargs)
+        elements.append(finat_elem)
+        deps.update(ds)
+    return finat.TensorProductElement(elements), deps
```

Comment:
Would be useful to explain these `shift_axes` and so forth a little.  Perhaps it is clearer when one sees the finat code?

Reply:
IIRC it was something like if you have a TPE(P4, P4, P4), then each P4 element is translated differently.